### PR TITLE
New resource & data source: aws_grafana_workspace

### LINKED ISSUE #16789: Add support for Amazon Managed Service for Grafana
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon has added a preview release of a managed Grafana service. Many users who are currently using EKS are using self-hosted Grafana instances for metrics dashboards.

### References

https://aws.amazon.com/grafana/

* #16788 - Add support for Amazon Managed Service for Prometheus

### PR DESCRIPTION
**Pull request by AlexanderSchiff**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #16789

Output from acceptance testing:

```
$ make testacc TESTS=TestAccXXX PKG=grafana

...
```

### CODE REVIEW

**Comment by AlexanderSchiff** [comment]
I need some help with my tests:

```
=== RUN   TestAccGrafanaWorkspace_basic
=== PAUSE TestAccGrafanaWorkspace_basic
=== CONT  TestAccGrafanaWorkspace_basic
    workspace_test.go:20: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # aws_grafana_workspace.test will be updated in-place
          ~ resource "aws_grafana_workspace" "test" {
                id                        = "g-545541a952"
                name                      = "tf-acc-test-3194149265356548228"
              - saml_configuration_status = "NOT_CONFIGURED" -> null
              - status                    = "ACTIVE" -> null
                # (4 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccGrafanaWorkspace_basic (154.84s)

FAIL
```

File: internal/service/grafana/find.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package grafana
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/managedgrafana"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
```
[the remark this answers is not in the document]
Comment (AlexanderSchiff) [reply internal/service/grafana/find.go]: Made that change. I am still getting an error on my tests, I am not sure what is wrong:
```
=== RUN   TestAccGrafanaWorkspace_basic
=== PAUSE TestAccGrafanaWorkspace_basic
=== CONT  TestAccGrafanaWorkspace_basic
=== CONT  TestAccGrafanaWorkspace_basic
    workspace_test.go:20: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # aws_grafana_workspace.test will be updated in-place
          ~ resource "aws_grafana_workspace" "test" {
                id                        = "g-717dfc8e45"
                name                      = "tf-acc-test-7310005638429688407"
              - saml_configuration_status = "NOT_CONFIGURED" -> null
              - status                    = "ACTIVE" -> null
                # (4 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccGrafanaWorkspace_basic (165.66s)

```

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TESTS=TestAccGrafana_serial PKG=grafana
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/grafana/... -v -count 1 -parallel 20 -run='TestAccGrafana_serial'  -timeout 180m
=== RUN   TestAccGrafana_serial
=== RUN   TestAccGrafana_serial/Workspace
=== RUN   TestAccGrafana_serial/Workspace/dataSources
=== RUN   TestAccGrafana_serial/Workspace/permissionType
=== RUN   TestAccGrafana_serial/Workspace/notificationDestinations
=== RUN   TestAccGrafana_serial/Workspace/saml
=== RUN   TestAccGrafana_serial/Workspace/sso
=== RUN   TestAccGrafana_serial/Workspace/disappears
=== RUN   TestAccGrafana_serial/Workspace/organization
=== RUN   TestAccGrafana_serial/DataSource
=== RUN   TestAccGrafana_serial/DataSource/basic
--- PASS: TestAccGrafana_serial (1328.09s)
    --- PASS: TestAccGrafana_serial/Workspace (1182.73s)
        --- PASS: TestAccGrafana_serial/Workspace/dataSources (155.28s)
        --- PASS: TestAccGrafana_serial/Workspace/permissionType (225.56s)
        --- PASS: TestAccGrafana_serial/Workspace/notificationDestinations (169.48s)
        --- PASS: TestAccGrafana_serial/Workspace/saml (180.94s)
        --- PASS: TestAccGrafana_serial/Workspace/sso (154.15s)
        --- PASS: TestAccGrafana_serial/Workspace/disappears (145.36s)
        --- PASS: TestAccGrafana_serial/Workspace/organization (151.96s)
    --- PASS: TestAccGrafana_serial/DataSource (145.36s)
        --- PASS: TestAccGrafana_serial/DataSource/basic (145.36s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/grafana	1331.810s
```

##### GovCloud

```console
% make testacc TESTS=TestAccGrafana_serial PKG=grafana 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/grafana/... -v -count 1 -parallel 20 -run='TestAccGrafana_serial'  -timeout 180m
=== RUN   TestAccGrafana_serial
=== RUN   TestAccGrafana_serial/DataSource
=== RUN   TestAccGrafana_serial/DataSource/basic
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace
=== RUN   TestAccGrafana_serial/Workspace/disappears
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace/organization
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace/dataSources
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace/permissionType
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace/notificationDestinations
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace/saml
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
=== RUN   TestAccGrafana_serial/Workspace/sso
    acctest.go:610: skipping tests; partition aws-us-gov does not support grafana service
--- PASS: TestAccGrafana_serial (1.03s)
    --- PASS: TestAccGrafana_serial/DataSource (1.03s)
        --- SKIP: TestAccGrafana_serial/DataSource/basic (1.03s)
    --- PASS: TestAccGrafana_serial/Workspace (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/disappears (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/organization (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/dataSources (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/permissionType (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/notificationDestinations (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/saml (0.00s)
        --- SKIP: TestAccGrafana_serial/Workspace/sso (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/grafana	4.776s
```

**Comment by ewbankkit** [comment]
@AlexanderSchiff Thanks for the contribution :tada: :clap:.
Overall the PR was in great shape.
I made some minor changes:

* Removed `created_date`, `last_updated_date` and `status` from the resource but kept them in the data source
* Made `authentication_providers` ForceNew as updating workspace authentication to SAML requires a complex configuration - Maybe this can be a future additional resource
* Named the required argument for the data source `workspace_id`

**Pull request merged**